ros2/rcl_handle/include/rcl_handle/action.hpp

```diff
@@ -0,0 +1,167 @@
+using accepted_callback =
+    std::function<void(std::shared_ptr<interface::server_goal<T>> const)>;
+
+namespace interface {
+/// \brief Allows action server to be mocked
+template <typename T>
+struct action_server {
+  /// \brief Allows action server to be instantiated outside consumer and
+  ///        for the consumer to then pass in methods or lambdas as callbacks
+  ///        Example:
+  /// \code{.cpp}
+  /// // test.cpp
+  /// auto fake_server = std::make_unique<fake_action_server>();
+  /// // decorate or define behavior for fake server
+  /// auto consumer = consumer{std::move(fake_server)};
+  /// // consumer.cpp
+  /// consumer(std::unique_ptr<action_server<Message>> as)
+  ///     : as_{std::move(as)} {
+  ///       as_->register_handles(
+  ///       [](auto const&, auto) {
+  ///         return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+  ///       },
+  ///       [](auto) { return rclcpp_action::CancelResponse::ACCEPT; },
+  ///       std::bind(&Consumer::handle_accepted, this, _1)
+  ///       );
+  /// \endcode
+  virtual void register_handles(goal_callback<T>, cancel_callback<T>,
+                                accepted_callback<T>) = 0;
+  virtual ~action_server() = default;
+};
+}  // namespace interface
+
+/// \brief Wrapper for server goal
+/// \tparam T Action message type
+template <typename T>
+struct server_goal final : public interface ::server_goal<T> {
+  explicit server_goal(
+      std::shared_ptr<rclcpp_action::ServerGoalHandle<T>> handle)
+      : handle_{std::move(handle)} {}
+  void publish_feedback(
+      std::shared_ptr<typename T::Feedback> feedback) override {
+    return handle_->publish_feedback(std::move(feedback));
+  }
+  void abort(std::shared_ptr<typename T::Result> result) override {
+    handle_->abort(std::move(result));
+  }
+  void succeed(std::shared_ptr<typename T::Result> result) override {
+    handle_->succeed(std::move(result));
+  }
+  void canceled(std::shared_ptr<typename T::Result> result) override {
+    handle_->canceled(std::move(result));
+  }
+  void execute() override { handle_->execute(); }
+  std::shared_ptr<typename T::Goal const> const get_goal() const override {
+    return handle_->get_goal();
+  }
+  rclcpp_action::GoalUUID const &get_goal_id() const override {
+    return handle_->get_goal_id();
+  }
+  bool is_canceling() const override { return handle_->is_canceling(); }
+  bool is_active() const override { return handle_->is_active(); }
+  bool is_executing() const override { return handle_->is_executing(); }
+
+ private:
+  std::shared_ptr<rclcpp_action::ServerGoalHandle<T>> handle_;
+};
+
+/// \brief Wrapper for action server
+/// \tparam T Action message type
+template <typename T>
+struct action_server final : public interface::action_server<T> {
+  action_server(std::shared_ptr<rclcpp::Node> node, std::string topic)
+      : node_{std::move(node)}, topic_{std::move(topic)} {}
+
+  void register_handles(goal_callback<T> goal, cancel_callback<T> cancel,
```

Comment:
Just to make sure I'm following, the idea is to use the Type Erasure idiom to DI the action server callbacks, right?

Reply:
That will be but there is no type erasure in this PR